electrolytes/solvation_shell_extract.py

```diff
@@ -0,0 +1,232 @@
+        skip_solvent_centered_shells: Skip extracting solvent-centered shells.
+        max_frames: Maximum number of frames to read from the trajectory.
+        shells_per_frame: Number of solutes or solvents per MD simulation frame from which to extract candidate shells.
+        max_shell_size: Maximum size (in atoms) of saved shells.
+        top_n: Number of snapshots to extract per topology.
+    """
+
+    # Read a structure and metadata file
+    logging.info("Reading structure and metadata files")
+
+    # Read metadata
+    with open(os.path.join(input_dir, "metadata_system.json")) as f:
+        metadata = json.load(f)
+
+    validate_metadata_file(metadata)
+
+    partial_charges = np.array(metadata["partial_charges"])
+
+    solutes = {}
+    solvents = {}
+
+    for res, species, spec_type in zip(
+        metadata["residue"], metadata["species"], metadata["solute_or_solvent"]
+    ):
+        if spec_type == "solute":
+            solutes[species] = res
+        elif spec_type == "solvent":
+            solvents[species] = res
+    spec_dicts = {"solute": solutes, "solvent": solvents}
+
+    # Read structures
+    structures = list(StructureReader(os.path.join(input_dir, "system_output.pdb")))
+    if max_frames > 0:
+        structures = random.sample(structures, max_frames)
+    # assign partial charges to atoms
+    logging.info("Assigning partial charges to atoms")
+    for st in tqdm(structures):
+        for at, charge in zip(st.atom, partial_charges):
+            at.partial_charge = charge
+
+    # For each solute: extract shells around the solute of some heuristic radii and bin by composition/graph hash
+    # Choose the N most diverse in each bin
+    spec_types = ["solute"]
+    if not skip_solvent_centered_shells:
+        spec_types.append("solvent")
+
+    for spec_type in spec_types:
+        for species, residue in spec_dicts[spec_type].items():
+            logging.info(f"Extracting solvation shells around {species}")
+            for radius in radii:
+                logging.info(f"Radius = {radius} A")
+                extracted_shells = []
+                for i, st in tqdm(
+                    enumerate(structures), total=len(structures)
+                ):  # loop over timesteps
+
+                    # extract shells from the structure, centered on the residue type
+                    extracted_shells.extend(
+                        extract_shells_from_structure(
+                            st,
+                            radius,
+                            residue,
+                            spec_type,
+                            shells_per_frame,
+                            max_shell_size,
+                        )
+                    )
+
+                if spec_type == "solvent":
+                    # raise a warning and continue to the next radii/species
+                    logging.warning("No solute-free shells found for solvent")
```

Comment:
Why are we always raising a warning here? It seems like we should first check the composition of `extracted_shells`. What am I missing?

Reply:
A casualty of me copying things around when I was refactoring.